### PR TITLE
Watch

### PR DESCRIPTION
**Pull request by BohdanBusko**
**STATUS**
READY

**DESCRIPTION**
add watch action for task

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [x] code coverage >90%
- [ ] DB indexes added if needed
- [x] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [x] I've checked new feature as logged in and logged out user if needed
- [x] "let!" is used in specs

### CODE REVIEW

File: app/controllers/account/tasks_controller.rb [hunk 1]
```diff
@@ -36,12 +38,24 @@ def update
...
-    resource.update_attributes(task_movement_params)
-    redirect_to account_workspace_project_path(parent.workspace_id, parent.id)
+    resource.update(task_movement_params)
+    redirect_to account_workspace_project_path(parent.workspace_id, parent)
+  end
+
+  def watch
+    @task = resource
+
+    if @task.users.include? current_user
```
Comment (igorkasyanchuk) [inline app/controllers/account/tasks_controller.rb]: wrap include? in "( )"

File: app/controllers/account/tasks_controller.rb [hunk 2]
```diff
@@ -36,12 +38,24 @@ def update
...
+      @task.watches.find_by(user_id: current_user).destroy
+    else
+      @task.update(users: [current_user])
```
Comment (igorkasyanchuk) [inline app/controllers/account/tasks_controller.rb]: this is not good. If you have 5 followers already - this code will delete all others I guess ... check it

File: app/assets/javascripts/application.js [hunk 3]
```diff
@@ -10,7 +10,7 @@
 // Read Sprockets README (https://github.com/rails/sprockets#sprockets-directives) for details
 // about supported directives.
 //
-//= require rails-ujs
+//= require jquery3
+//= require jquery_ujs
```
Comment (igorkasyanchuk) [inline app/assets/javascripts/application.js:14]: get back turbolinks

File: app/models/task.rb [hunk 4]
```diff
@@ -27,6 +27,8 @@ class Task < ApplicationRecord
   ranks :row_order, with_same: :project_id
 
   belongs_to :project, required: true
+  has_many :watches, dependent: :destroy
+  has_many :users, through: :watches
```
Comment (igorkasyanchuk) [inline app/models/task.rb]: not "users". rename to "watchers"

and rename Watch to TaskWatch

File: app/views/account/tasks/show.html.slim [hunk 5]
```diff
@@ -1,8 +1,14 @@
...
 .container
   .content
-    .title
-      = "#{@project.name}: #{@task.title}"
+    .level
+      .level-left
+        .title
+          = "#{@project.name}: #{@task.title}"
+      .level-right
+        = link_to (@task.users.include?(current_user) ? 'Following' : 'Follow task'),
```
Comment (igorkasyanchuk) [inline app/views/account/tasks/show.html.slim]: create method @user.watching?(@task)

File: config/routes.rb [hunk 6]
```diff
@@ -15,6 +15,7 @@
       resources :tasks, except: [:index] do
         member do
           put :move
+          get :watch
```
Comment (igorkasyanchuk) [inline config/routes.rb]: why get? 

you creating something in DB, so it's POST

File: db/migrate/20181025132424_add_section_to_task.rb [hunk 7]
```diff
@@ -1,5 +1,5 @@
 class AddSectionToTask < ActiveRecord::Migration[5.2]
   def change
-    add_column :tasks, :section, :boolean, default: false
+    add_column :tasks, :section, :boolean
```
Comment (igorkasyanchuk) [inline db/migrate/20181025132424_add_section_to_task.rb]: where is default value for column?

File: app/controllers/account/tasks_controller.rb (hunk 2 above)
Comment (igorkasyanchuk) [reply app/controllers/account/tasks_controller.rb]: I expect code something like

@task.add_watcher(current_user)
@task.remove_watcher(current_user)

File: spec/models/watch_spec.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+#
+# Foreign Keys
+#
+#  fk_rails_...  (task_id => tasks.id)
+#  fk_rails_...  (user_id => users.id)
+#
+
+require 'rails_helper'
+
+RSpec.describe Watch, type: :model do
```
Comment (igorkasyanchuk) [inline spec/models/watch_spec.rb]: wheere is spec for this model? and test to see if you have factory

File: app/controllers/account/tasks_controller.rb (hunk 1 above)
Comment (romaslmd) [reply app/controllers/account/tasks_controller.rb]: no need to use for class lvl methods: `attr_reader` `validate` that u usually call on class lvl
but for others - stick to the point of readability

File: app/views/account/tasks/show.html.slim [hunk 9]
```diff
@@ -1,8 +1,14 @@
 - title "#{@project.name}: #{@task.title}"
 .container
   .content
-    .title
-      = "#{@project.name}: #{@task.title}"
+    .level
+      .level-left
+        .title
```
Comment (igorkasyanchuk) [inline app/views/account/tasks/show.html.slim]: where is level-item?

File: app/models/task.rb [hunk 10]
```diff
@@ -27,9 +27,19 @@ class Task < ApplicationRecord
...
 
   validates :title, length: { maximum: 250 }, presence: true
   validates :description, length: { maximum: 250 }
+
+  def add_watcher(user)
+    self.update(watchers: [user])
+  end
+
+  def remove_watcher(user)
+    self.task_watches.find_by(user_id: user, task_id: self).destroy
```
Comment (igorkasyanchuk) [inline app/models/task.rb]: 1. self.task_watches.find_by(user_id: user, task_id: self) не треба , task_id: self....там вже є таск
2. що буде якщо нема вотчера. додай перевірку чи не nil перед destroy

File: app/models/user.rb [hunk 11]
```diff
@@ -63,4 +65,8 @@ def self.new_with_session(params, session)
       end
     end
   end
+
+  def watching?(task)
+    self.tasks.include?(task)
```
Comment (igorkasyanchuk) [inline app/models/user.rb]: What SQL query it generates?
change to active record. Does it return all tasks as relation?

use "where(...).any?"

File: spec/controllers/account/tasks_controller_spec.rb [hunk 12]
```diff
@@ -6,52 +6,80 @@
...
 
       expect(response).to redirect_to account_workspace_project_path(workspace.id, project.id)
-      expect(project.tasks.row_order_asc.fourth).to eq project.tasks.order(id: :asc).last
+      expect(project.tasks.row_order_asc).to eq [task1, task3, task2]
+    end
+  end
+
+  context 'GET /watch' do
+    it 'should follow task' do
+      get :watch, params: { project_id: project.id, id: task1.id }, format: :js
```
Comment (igorkasyanchuk) [inline spec/controllers/account/tasks_controller_spec.rb]: додай тест такий
1. передумова - є таск в якого є 2 вотчера
2. інший юзер пробує стати теж вотчером
3. перевіряєш що у таска тепер 3 вотчера

File: app/controllers/account/tasks_controller.rb [hunk 13]
```diff
@@ -1,4 +1,6 @@
 class Account::TasksController < Account::AccountController
+  protect_from_forgery except: :watch
```
Comment (romaslmd) [inline app/controllers/account/tasks_controller.rb]: eh?

File: app/models/user.rb [hunk 14]
```diff
@@ -63,4 +65,8 @@ def self.new_with_session(params, session)
       end
     end
   end
+
+  def watching?(task)
+    self.tasks.where(id: task).any?
```
Comment (romaslmd) [inline app/models/user.rb]: prefer `#exists?` over `#any?`

Comment (romaslmd) [reply app/models/user.rb]: btw, better to pass `.where(id: task.id)`

File: app/views/account/tasks/show.html.slim [hunk 15]
```diff
@@ -1,8 +1,17 @@
...
-    .title
-      = "#{@project.name}: #{@task.title}"
+    .level
+      .level-left
+        .level-item
+          .title
+            = "#{@project.name}: #{@task.title}"
+      .level-right
+        .level-item
+          = link_to (current_user.watching?(@task) ? 'Following' : 'Follow task'),
```
Comment (romaslmd) [inline app/views/account/tasks/show.html.slim]: helper?

File: app/views/account/tasks/show.html.slim [hunk 16]
```diff
@@ -1,8 +1,17 @@
...
+            watch_account_project_task_path(@project, @task),
+            class: 'button is-text', remote: true, id: "watch_task_#{@task.id}",
```
Comment (romaslmd) [inline app/views/account/tasks/show.html.slim]: use `#dom_id` helper for dynamic ids

File: app/views/account/tasks/watch.js.erb [hunk 17]
```diff
@@ -0,0 +1 @@
+$("#watch_task_<%=@task.id%>").html("<%= current_user.watching?(@task) ? 'Following' : 'Follow task' %>");
```
Comment (romaslmd) [inline app/views/account/tasks/watch.js.erb]: move the link to a partial and re-render it here

File: config/routes.rb [hunk 18]
```diff
@@ -15,6 +15,7 @@
       resources :tasks, except: [:index] do
         member do
           put :move
+          post :watch
```
Comment (romaslmd) [inline config/routes.rb]: why POST? `PATCH` sounds more relevant

File: db/migrate/20181030180602_change_column_default.rb [hunk 19]
```diff
@@ -0,0 +1,5 @@
+class ChangeColumnDefault < ActiveRecord::Migration[5.2]
+  def change
+    change_column_default :tasks, :section, false
```
Comment (romaslmd) [inline db/migrate/20181030180602_change_column_default.rb]: no need to create a separate migration to edit one you're creating in a one propagation

File: app/models/task.rb [hunk 20]
```diff
@@ -27,9 +27,19 @@ class Task < ApplicationRecord
...
+  has_many :task_watches, dependent: :destroy
+  has_many :watchers, through: :task_watches, source: :user
 
   scope :row_order_asc, -> { order(row_order: :asc) }
 
   validates :title, length: { maximum: 250 }, presence: true
   validates :description, length: { maximum: 250 }
+
+  def add_watcher(user)
+    self.task_watches.create(user_id: user.id)
```
Comment (romaslmd) [inline app/models/task.rb]: prefer `find_or_create_by` to avoid duplications

File: app/models/task.rb [hunk 21]
```diff
@@ -27,9 +27,19 @@ class Task < ApplicationRecord
...
+  end
+
+  def remove_watcher(user)
+    self.task_watches.find_by(user_id: user).destroy if user
```
Comment (romaslmd) [inline app/models/task.rb]: better to replace with
```
self.task_watches.where(user: user).destroy_all
```

NOTE: `find_by` may return `nil`

File: app/views/account/tasks/show.html.slim (hunk 15 above)
Comment (BohdanBusko) [reply app/views/account/tasks/show.html.slim]: No it`s the model's method. Or you mean to add helper?

File: config/routes.rb (hunk 18 above)
Comment (BohdanBusko) [reply config/routes.rb]: it creates not update

File: app/views/account/tasks/watch.js.erb (hunk 17 above)
Comment (BohdanBusko) [reply app/views/account/tasks/watch.js.erb]: а хіба не простіше просто міняти текст в посиланні ніж перерендерювати ціле посилання?

File: app/views/account/tasks/show.html.slim (hunk 16 above)
Comment (BohdanBusko) [reply app/views/account/tasks/show.html.slim]: А як мені потім взяти потрібний мені елемент? Якщо в нього динамічне id?

File: db/migrate/20181030180602_change_column_default.rb (hunk 19 above)
Comment (BohdanBusko) [reply db/migrate/20181030180602_change_column_default.rb]: в мене була бага тому прийшлося зробити міграцію

File: app/models/task.rb (hunk 21 above)
Comment (BogdanBusko) [reply app/models/task.rb]: В контроллері є перевірка

File: app/views/account/tasks/show.html.slim (hunk 15 above)
Comment (romaslmd) [reply app/views/account/tasks/show.html.slim]: no, i meant - helper to render the link

File: app/views/account/tasks/show.html.slim (hunk 16 above)
Comment (romaslmd) [reply app/views/account/tasks/show.html.slim]: feel free to test how [it's](https://apidock.com/rails/ActionView/RecordIdentifier/dom_id) working

Comment (romaslmd) [reply app/views/account/tasks/show.html.slim]: btw, `"watch_task_#{@task.id}"` - this is already a dynamic id ;)

File: app/views/account/tasks/watch.js.erb (hunk 17 above)
Comment (romaslmd) [reply app/views/account/tasks/watch.js.erb]: nope, what if later you need to add/remove class, change icon etc.?

File: config/routes.rb (hunk 18 above)
Comment (romaslmd) [reply config/routes.rb]: really? from what i see you're within `:tasks` resource, and update task record by adding a watcher.. isn't that correct?

File: db/migrate/20181030180602_change_column_default.rb (hunk 19 above)
Comment (romaslmd) [reply db/migrate/20181030180602_change_column_default.rb]: more details?
seams like it wasn't a bug..

File: app/views/account/tasks/watch.js.erb [hunk 22]
```diff
@@ -0,0 +1,2 @@
+$("#watch_task_<%=@task.id%> a").html('<%= j render 'watch_link'%>');
```
Comment (igorkasyanchuk) [inline app/views/account/tasks/watch.js.erb]: add spaces arround "@task.id" code must be more readable

File: app/views/account/tasks/show.html.slim [hunk 23]
```diff
@@ -1,8 +1,13 @@
...
   .content
-    .title
-      = "#{@project.name}: #{@task.title}"
+    .level
+      .level-left
+        .level-item
+          .title
+            = "#{@project.name}: #{@task.title}"
+      .level-right
+        = content_tag(:div, render('watch_link'), class: 'level-item', id: "watch_task_#{@task.id}")
```
Comment (igorkasyanchuk) [inline app/views/account/tasks/show.html.slim]: use slim
why content_tag?

Comment (igorkasyanchuk) [reply app/views/account/tasks/show.html.slim]: and put full path to partial

File: app/views/account/tasks/watch.js.erb (hunk 22 above)
Comment (igorkasyanchuk) [reply app/views/account/tasks/watch.js.erb]: always put full path to partial

File: app/views/account/tasks/_watch_link.html.slim [hunk 24]
```diff
@@ -0,0 +1,3 @@
+= link_to (current_user.watching?(@task) ? 'Following' : 'Follow task'),
```
Comment (igorkasyanchuk) [inline app/views/account/tasks/_watch_link.html.slim]: don;t pass instance vairables

**Review by igorkasyanchuk — APPROVED** [review]

**Pull request merged**